cura/API/Machines.py

```diff
@@ -0,0 +1,139 @@
+            group_id = global_stack.getMetaDataEntry("group_id")
+            machine_manager = self._application.getMachineManager()
+            for machine in machine_manager.getMachinesInGroup(group_id):
+                machine.setMetaDataEntry("group_name", group_name)
+
+            # Set the default value for "hidden", which is used when you have a group with multiple types of printers
+            global_stack.setMetaDataEntry("hidden", False)
+
+    ##  Set the current machine's configuration from an (optional) output device.
+    #   If no output device is given, the first one available on the machine will be used.
+    #   NOTE: Group and machine are used interchangeably.
+    #   NOTE: This doesn't seem to be used anywhere. Maybe delete?
+    @pyqtSlot(QObject)
+    def updateCurrentMachineConfiguration(self, output_device: Optional["PrinterOutputDevice"]) -> None:
+
+        if output_device is None:
+            machine_manager = self._application.getMachineManager()
+            output_device = machine_manager.printerOutputDevices[0]
+        
+        hotend_ids = output_device.hotendIds
+        for index in range(len(hotend_ids)):
+            output_device.hotendIdChanged.emit(index, hotend_ids[index])
+        
+        material_ids = output_device.materialIds
+        for index in range(len(material_ids)):
+            output_device.materialIdChanged.emit(index, material_ids[index])
+
+    ##  Add an output device to the current machine.
+    #   In practice, this means:
+    #   - Setting the output device's network key in the current machine's metadata
+    #   - Adding the output device's connection type to the current machine's configured connection
+    #     types.
+    #   TODO: CHANGE TO HOSTNAME
+    @pyqtSlot(QObject)
+    def addOutputDeviceToCurrentMachine(self, output_device: "PrinterOutputDevice") -> None:
+        if not output_device:
+            return
+        Logger.log("d",
+            "Attempting to set the network key of the active machine to %s",
+            output_device.key)
+        global_stack = self._application.getGlobalContainerStack()
+        if not global_stack:
+            return
+        metadata = global_stack.getMetaData()
+
+        # Global stack already had a connection, but it's changed.
+        if "um_network_key" in metadata: 
+            old_network_key = metadata["um_network_key"]
+
+            # Since we might have a bunch of hidden stacks, we also need to change it there.
+            metadata_filter = {"um_network_key": old_network_key}
+            containers = self._application.getContainerRegistry().findContainerStacks(
+                type = "machine", **metadata_filter)
+            for container in containers:
+                container.setMetaDataEntry("um_network_key", output_device.key)
+
+                # Delete old authentication data.
+                Logger.log("d", "Removing old authentication id %s for device %s",
```

Comment:
`"i"`?

Reply:
`"d"` fits with the other debug statements. It's the sort of thing you only really care about when debugging.

Reply:
cool